File: leetcode/src/00062_unique_paths/solution.go
package leetcode

func uniquePaths(m int, n int) int {
	dp := [2][]int{}
	for i := 0; i < 2; i++ {
		dp[i] = make([]int, n)
		for j := 0; j < n; j++ {
			dp[i][j] = 1
		}
	}
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			dp[i&1][j] = dp[(i-1)&1][j] + dp[i&1][j-1]
		}
	}
	return dp[(m-1)&1][n-1]
}

File: leetcode/src/00062_unique_paths/solution.rs
struct Solution;

impl Solution {
    pub fn unique_paths(m: i32, n: i32) -> i32 {
        let n = n as usize;
        let mut dp = vec![1; n];
        for _ in 1..m {
            for j in 1..n {
                dp[j] += dp[j - 1];
            }
        }
        dp[n - 1]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let m = 3;
        let n = 7;
        assert_eq!(Solution::unique_paths(m, n), 28);

        let m = 3;
        let n = 2;
        assert_eq!(Solution::unique_paths(m, n), 3);
    }
}
